### PR TITLE
Add parsing support for EIP 2930 transactions

### LINKED ISSUE #6494: Add EIP 2930 transaction parsing support to importer
### Problem

[Services Issues 7670](https://github.com/hashgraph/hedera-services/issues/7670) highlighted that the Hedera does not yet support EIP 2930 types transactions.

To support the E2E flow the mirror node parser should be able to extract detailed from an EIP 2930 RLP encoded transaction.

### Solution

Add support to the Mirror Node parser to handle EIP 2930 transactions
Suggestion
- Add Eip2930EthereumTransactionParser
- Add Eip2930EthereumTransactionParserTest for coverage
- Update `CompositeEthereumTransactionParser`
  - Add `Eip2930EthereumTransactionParser` as a parser option
  - Update `getEthereumTransactionParser()` to manage the type 1 option

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Nana-EC**
**Description**:
Add support for EIP 2930 (type 1) transactions

- Add Eip2930EthereumTransactionParser
- Add Eip2930EthereumTransactionParserTest for coverage
- Update CompositeEthereumTransactionParser
  - Add Eip2930EthereumTransactionParser as a parser option
  - Update getEthereumTransactionParser() to manage the type 1 option

**Related issue(s)**:

Fixes #6494 

**Notes for reviewer**:
Basic parsing and testing logic
Further consideration for 
- [ ] Impacted tests
- [ ] Integration test
- [ ] acceptance test (if applicable)

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip1559EthereumTransactionParser.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 @Named
 public class Eip1559EthereumTransactionParser implements EthereumTransactionParser {
     public static final int EIP1559_TYPE_BYTE = 2;
+    public static final String EIP1559_TYPE_HEX = "0x02";
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip1559EthereumTransactionParser.java:27]: Seems to be unused so should revert.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.esaulpaugh.headlong.rlp.RLPDecoder;
+import com.hedera.mirror.common.domain.transaction.EthereumTransaction;
+import com.hedera.mirror.importer.exception.InvalidEthereumBytesException;
+import jakarta.inject.Named;
+
+@Named
+public class Eip2930EthereumTransactionParser implements EthereumTransactionParser {
+    public static final int EIP2930_TYPE_BYTE = 1;
+    public static final String EIP2930_TYPE_HEX = "0x01";
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java:27]: This field seems to be unused

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public EthereumTransaction decode(byte[] transactionBytes) {
+        var decoder = RLPDecoder.RLP_STRICT.sequenceIterator(transactionBytes);
+        var legacyRlpItem = decoder.next();
+        var legacyRlpItemByte = legacyRlpItem.asByte();
+        if (legacyRlpItemByte != EIP2930_TYPE_BYTE) {
+            throw new InvalidEthereumBytesException(
+                    TRANSACTION_TYPE_NAME,
+                    String.format("1st byte was %s but should " + "be %s", legacyRlpItemByte, EIP2930_TYPE_BYTE));
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java:39]:
```suggestion
                    String.format("First byte was %s but should be %s", legacyRlpItemByte, EIP2930_TYPE_BYTE));
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public static final String EIP2930_TYPE_HEX = "0x01";
+    private static final String TRANSACTION_TYPE_NAME = "EIP2930";
+    private static final int EIP2930_TYPE_RLP_ITEM_COUNT = 11;
+
+    @Override
+    public EthereumTransaction decode(byte[] transactionBytes) {
+        var decoder = RLPDecoder.RLP_STRICT.sequenceIterator(transactionBytes);
+        var legacyRlpItem = decoder.next();
+        var legacyRlpItemByte = legacyRlpItem.asByte();
+        if (legacyRlpItemByte != EIP2930_TYPE_BYTE) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java:36]: Why is this called a legacyRlpItem?

```suggestion
        var type = decoder.next();
        var typeByte = type.asByte();
        if (typeByte != EIP2930_TYPE_BYTE) {
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        if (!eip2930RlpItem.isList()) {
+            throw new InvalidEthereumBytesException(TRANSACTION_TYPE_NAME, "2nd RLPItem was not a list");
+        }
+
+        var rlpItems = eip2930RlpItem.asRLPList().elements();
+        if (rlpItems.size() != EIP2930_TYPE_RLP_ITEM_COUNT) {
+            throw new InvalidEthereumBytesException(
+                    TRANSACTION_TYPE_NAME,
+                    String.format(
+                            "2nd RLPItem list size was " + "%s but should be %s",
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java:52]:
```suggestion
                            "RLP list size was %d but expected %d",
```

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                .hasMessage("Unable to decode EIP2930 ethereum transaction bytes, 2nd RLPItem list size was 0 but "
+                        + "should be 11");
+    }
+
+    @Override
+    protected void validateEthereumTransaction(EthereumTransaction ethereumTransaction) {
+        assertThat(ethereumTransaction)
+                .isNotNull()
+                .satisfies(t -> assertThat(t.getType()).isEqualTo(Eip2930EthereumTransactionParser.EIP2930_TYPE_BYTE))
+                .satisfies(t -> assertThat(Hex.encodeHexString(t.getChainId())).isEqualTo("012a"))
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java:88]: Most if not all can be converted to the simpler returns:

```suggestion
                .returns(EIP2930_TYPE_BYTE, EthereumTransaction::getType)
                .returns("012a", EthereumTransaction::getChainId)
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @SneakyThrows
+    @Override
+    protected void validateEthereumTransaction(EthereumTransaction ethereumTransaction) {
+        assertThat(ethereumTransaction)
+                .isNotNull()
+                .returns(Eip2930EthereumTransactionParser.EIP2930_TYPE_BYTE, EthereumTransaction::getType)
+                .returns(Hex.decodeHex("012a"), EthereumTransaction::getChainId)
+                .returns(0L, EthereumTransaction::getNonce)
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java:88]: Preferably we should use a transactions with a valid nonce.
I used available bytes at the time but may craft a transaction with > 0 nonce.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2022-2023 Hedera Hashgraph, LLC
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParserTest.java:2]: Not: I missed this in copying but Should probably be 2023

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2022-2023 Hedera Hashgraph, LLC
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java:2]: nit:2023

**Review by Nana-EC — COMMENTED** [review]
LG.
One test modification and nits on this I hadn't gotten to.



Additionally, we should have a REST API test that confirms this transactions types response.
Potentially not needed in this PR and can be in a follow up soon after.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip1559EthereumTransactionParser.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip1559EthereumTransactionParser.java]: Yeah left over from an initial approach I considered. Can remove.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/Eip2930EthereumTransactionParser.java]: Yeah left over from an initial approach I considered. Can remove.

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**